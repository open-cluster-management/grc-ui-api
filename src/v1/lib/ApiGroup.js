/* Copyright (c) 2020 Red Hat, Inc. */
// acm api group and version
const ApiGroup = {
  hostUrl: 'http://0.0.0.0/kubernetes/apis',
  policiesGroup: 'policy.open-cluster-management.io',
  appsGroup: 'apps.open-cluster-management.io',
  version: 'v1',
  clusterInfoGroup: 'internal.open-cluster-management.io',
  clusterAPIVersion: 'v1beta1',
};

export default ApiGroup;
