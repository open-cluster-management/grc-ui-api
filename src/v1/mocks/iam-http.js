/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

export default function createMockHttp() {
  const state = {
    namespaces: {
      items: [
        { // non cluster namespace
          metadata: {
            name: 'default',
          },
        },
        { // non cluster namespace
          metadata: {
            name: 'kube-system',
          },
        },
        { // non cluster namespace
          metadata: {
            name: 'policy-namespace',
          },
        },
        { // cluster namespace
          metadata: {
            name: 'cluster1',
          },
        },
        { // cluster namespace
          metadata: {
            name: 'local-cluster',
          },
        },
      ],
    },
  };

  return async function MockLib(params) {
    switch (true) {
      case params.url.includes('resourceType=namespace'):
      default:
        return state.namespaces;
    }
  };
}
