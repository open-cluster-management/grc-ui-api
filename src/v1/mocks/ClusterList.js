/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
/* Copyright (c) 2020 Red Hat, Inc. */

import ApiGroup from '../lib/ApiGroup';

export const mockCluster1Response = {
  kind: 'ClusterList',
  apiVersion: 'clusterregistry.k8s.io/v1alpha1',
  metadata: {
    selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/cluster1/clusters/`,
    resourceVersion: '219916',
  },
  items: [
    {
      kind: 'Cluster',
      apiVersion: 'clusterregistry.k8s.io/v1alpha1',
      metadata: {
        name: 'cluster1',
        namespace: 'cluster1',
        selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/cluster1/clusters/cluster1`,
        uid: '6120d6bb-e482-11e9-aa00-96bb9600cc18',
        resourceVersion: '219786',
        creationTimestamp: '2019-10-01T19:33:43Z',
        labels: {
          environment: 'Dev',
          name: 'cluster1',
          region: 'US',
        },
        annotations: {
          'mcm.ibm.com/hub': 'hub0',
          'mcm.ibm.com/user-group': 'c3lzdGVtOnNlcnZpY2VhY2NvdW50cyxzeXN0ZW06c2VydmljZWFjY291bnRzOmt1YmUtc3lzdGVtLHN5c3RlbTphdXRoZW50aWNhdGVk',
          'mcm.ibm.com/user-identity': 'c3lzdGVtOnNlcnZpY2VhY2NvdW50Omt1YmUtc3lzdGVtOmRlZmF1bHQ=',
          'seed-generation': '2',
        },
        finalizers: [
          'platform-api.cluster',
          'propagator.finalizer.mcm.ibm.com',
        ],
      },
      spec: {
        kubernetesApiEndpoints: {
          serverEndpoints: [
            {
              clientCIDR: '0.0.0.0/0',
              serverAddress: '9.42.82.160:8001',
            },
          ],
        },
        authInfo: {

        },
      },
      status: {
        conditions: [
          {
            type: '',
            status: '',
            lastHeartbeatTime: '2019-10-31T17:10:33Z',
            lastTransitionTime: '2019-10-31T17:11:34Z',
            reason: 'Klusterlet failed to update cluster status on time',
          },
        ],
      },
    },
  ],
};

export const mockClusterIronmanResponse
 = [
   {
     apiVersion: 'internal.open-cluster-management.io/v1beta1',
     kind: 'ManagedClusterInfo',
     metadata: {
       creationTimestamp: '2020-06-16T13:58:50Z',
       generation: 2,
       labels: {
         cloud: 'auto-detect',
         environment: '',
         name: 'ironman',
         vendor: 'auto-detect',
       },
       name: 'ironman',
       namespace: 'ironman',
       resourceVersion: '5002550',
       selfLink: '/apis/internal.open-cluster-management.io/v1beta1/namespaces/ironman/managedclusterinfos/ironman',
       uid: '23b04acb-6d80-40c5-9db8-4a53782cbfb8',
     },
     spec: {
       masterEndpoint: 'api.ironman.dev08.red-chesterfield.com:6443',
     },
     status: {
       conditions: [
         {
           lastTransitionTime: '2020-06-19T13:13:59Z',
           message: 'cluster is posting ready status',
           reason: 'ClusterReady',
           status: 'True',
           type: 'ManagedClusterJoined',
         },
       ],
       distributionInfo: {
         ocp: {},
       },
       loggingEndpoint: {
         ip: '',
       },
       loggingPort: {
         port: 0,
       },
     },
   },
 ];

export const mockClusterDefaultResponse
 = [
   {
     apiVersion: 'internal.open-cluster-management.io/v1beta1',
     kind: 'ManagedClusterInfo',
     metadata: {
       creationTimestamp: '2020-06-16T13:58:50Z',
       generation: 2,
       labels: {
         cloud: 'auto-detect',
         environment: '',
         name: 'default',
         vendor: 'auto-detect',
       },
       name: 'default',
       namespace: 'default',
       resourceVersion: '5002550',
       selfLink: '/apis/internal.open-cluster-management.io/v1beta1/namespaces/default/managedclusterinfos/default',
       uid: '23b04acb-6d80-40c5-9db8-4a53782cbfb8',
     },
     spec: {
       masterEndpoint: 'api.default.dev08.red-chesterfield.com:6443',
     },
     status: {
       conditions: [
         {
           lastTransitionTime: '2020-06-19T13:13:59Z',
           message: 'cluster is posting ready status',
           reason: 'ClusterReady',
           status: 'True',
           type: 'ManagedClusterJoined',
         },
       ],
       distributionInfo: {
         ocp: {},
       },
       loggingEndpoint: {
         ip: '',
       },
       loggingPort: {
         port: 0,
       },
     },
   },
 ];

export const mockClusterKubeSystemResponse
 = [
   {
     apiVersion: 'internal.open-cluster-management.io/v1beta1',
     kind: 'ManagedClusterInfo',
     metadata: {
       creationTimestamp: '2020-06-16T13:58:50Z',
       generation: 2,
       labels: {
         cloud: 'auto-detect',
         environment: '',
         name: 'kube-system',
         vendor: 'auto-detect',
       },
       name: 'kube-system',
       namespace: 'kube-system',
       resourceVersion: '5002550',
       selfLink: '/apis/internal.open-cluster-management.io/v1beta1/namespaces/kube-system/managedclusterinfos/kube-system',
       uid: '23b04acb-6d80-40c5-9db8-4a53782cbfb8',
     },
     spec: {
       masterEndpoint: 'api.kube-system.dev08.red-chesterfield.com:6443',
     },
     status: {
       conditions: [
         {
           lastTransitionTime: '2020-06-19T13:13:59Z',
           message: 'cluster is posting ready status',
           reason: 'ClusterReady',
           status: 'True',
           type: 'ManagedClusterJoined',
         },
       ],
       distributionInfo: {
         ocp: {},
       },
       loggingEndpoint: {
         ip: '',
       },
       loggingPort: {
         port: 0,
       },
     },
   },
 ];

export const mockClusterHubResponse = {
  kind: 'ClusterList',
  apiVersion: 'clusterregistry.k8s.io/v1alpha1',
  metadata: {
    selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/clusterhub/clusters/`,
    resourceVersion: '219917',
  },
  items: [
    {
      kind: 'Cluster',
      apiVersion: 'clusterregistry.k8s.io/v1alpha1',
      metadata: {
        name: '',
        namespace: '',
        selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/clusterhub/clusters/clusterhub`,
        uid: '24466f0f-e482-11e9-aa00-96bb9600cc18',
        resourceVersion: '219917',
        creationTimestamp: '2019-10-01T19:32:01Z',
        labels: {
          environment: 'Dev',
          name: 'clusterhub',
          region: 'US',
        },
        annotations: {
          'mcm.ibm.com/hub': 'hub0',
          'mcm.ibm.com/user-group': 'aGNtOmNsdXN0ZXJzLHN5c3RlbTphdXRoZW50aWNhdGVk',
          'mcm.ibm.com/user-identity': 'aGNtOmNsdXN0ZXJzOmNsdXN0ZXJodWI6Y2x1c3Rlcmh1Yg==',
          'seed-generation': '2',
        },
        finalizers: [
          'platform-api.cluster',
          'propagator.finalizer.mcm.ibm.com',
        ],
      },
      spec: {
        kubernetesApiEndpoints: {
          serverEndpoints: [
            {
              clientCIDR: '0.0.0.0/0',
              serverAddress: '9.42.82.240:8001',
            },
          ],
        },
        authInfo: {

        },
      },
      status: {
        conditions: [
          {
            type: 'OK',
            status: '',
            lastHeartbeatTime: '2019-10-31T18:02:41Z',
            lastTransitionTime: '2019-10-19T21:47:19Z',
          },
        ],
      },
    },
  ],
};


export const mockMCMResponse = {
  kind: 'ClusterList',
  apiVersion: 'clusterregistry.k8s.io/v1alpha1',
  metadata: {
    selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/mcm/clusters/`,
    resourceVersion: '219919',
  },
  items: [],
};

export const mockDefaultResponse = {
  kind: 'ClusterList',
  apiVersion: 'clusterregistry.k8s.io/v1alpha1',
  metadata: {
    selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/default/clusters/`,
    resourceVersion: '219920',
  },
  items: [],
};

export const mockKubeSystemResponse = {
  kind: 'ClusterList',
  apiVersion: 'clusterregistry.k8s.io/v1alpha1',
  metadata: {
    selfLink: `/apis/${ApiGroup.clusterRegistryGroup}/${ApiGroup.mcmVersion}/namespaces/kube-system/clusters/`,
    resourceVersion: '219921',
  },
  items: [],
};
