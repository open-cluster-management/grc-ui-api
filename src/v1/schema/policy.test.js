/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019, 2020. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
/* Copyright (c) 2020 Red Hat, Inc. */

import supertest from 'supertest';
import nock from 'nock';
import server, { GRAPHQL_PATH } from '../index';
import ApiGroup from '../lib/ApiGroup';
import {
  mockPolicyListResponse, mockSinglePolicyResponse, mockCreatePolicy, mockDeleteResponse,
  mockClusterListResponse, mockCluster1ListResponse, mockClusterHubListResponse,
  mockDefaultListResponse, mockKubeSystemListResponse, mockViolationListResponse,
  mockNewAPISinglePolicyResponse,
} from '../mocks/PolicyList';
import { mockCluster1Response, mockClusterHubResponse, mockMCMResponse, mockDefaultResponse, mockKubeSystemResponse } from '../mocks/ClusterList';

describe('Policy Resolver', () => {
  beforeAll(() => {
    // specify the url to be intercepted
    const APIServer = nock(ApiGroup.hostUrl);

    APIServer.get(`/${ApiGroup.policiesGroup}/${ApiGroup.version}/namespaces/mcm/policies`)
      .reply(200, mockPolicyListResponse);

    APIServer.get(`/${ApiGroup.policiesGroup}/${ApiGroup.version}/namespaces/mcm/policies/policy-all`)
      .reply(200, mockSinglePolicyResponse);

    APIServer.post(`/${ApiGroup.policiesGroup}/${ApiGroup.version}/namespaces/mcm/policies`)
      .reply(200, mockCreatePolicy);

    APIServer.get('/mcm.ibm.com/v1alpha1/namespaces/mcm/clusterstatuses')
      .reply(200, mockClusterListResponse);

    APIServer.get('/mcm.ibm.com/v1alpha1/namespaces/cluster1/clusterstatuses')
      .reply(200, mockCluster1ListResponse);

    APIServer.get('/mcm.ibm.com/v1alpha1/namespaces/default/clusterstatuses')
      .reply(200, mockDefaultListResponse);

    APIServer.get('/mcm.ibm.com/v1alpha1/namespaces/kube-system/clusterstatuses')
      .reply(200, mockKubeSystemListResponse);

    APIServer.get('/mcm.ibm.com/v1alpha1/namespaces/clusterhub/clusterstatuses')
      .reply(200, mockClusterHubListResponse);

    APIServer.post('/policies.policies.open-cluster-management.io')
      .reply(200, mockViolationListResponse);

    APIServer.delete(`/${ApiGroup.policiesGroup}/${ApiGroup.version}/namespaces/default/policies/test-policy`)
      .reply(200, mockDeleteResponse);

    APIServer.delete('/clusterregistry.k8s.io/v1alpha1/namespaces/default/clusters')
      .reply(200, mockDefaultResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/default/clusters/cluster1')
      .reply(200, mockDefaultResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/default/clusters')
      .reply(200, mockDefaultResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/kube-system/clusters/cluster1')
      .reply(200, mockKubeSystemResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/kube-system/clusters')
      .reply(200, mockKubeSystemResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/mcm/clusters/cluster1')
      .reply(200, mockMCMResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/mcm/clusters')
      .reply(200, mockMCMResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/clusterhub/clusters/cluster1')
      .reply(200, mockDefaultResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/clusterhub/clusters')
      .reply(200, mockDefaultResponse);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/cluster1/clusters')
      .reply(200, mockDefaultResponse);

    // Single cluster
    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/cluster1/clusters/cluster1')
      .reply(200, mockCluster1Response);

    APIServer.get('/clusterregistry.k8s.io/v1alpha1/namespaces/clusterhub/clusters/clusterhub')
      .reply(200, mockClusterHubResponse);

    APIServer.get('/policies.open-cluster-management.io/v1/namespaces/calamari/policies/default.case1-test-policy')
      .reply(200, mockNewAPISinglePolicyResponse);
  });

  test('Correctly Resolves All Policies per Cluster List Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          policiesInCluster(cluster: "cluster1") {
            cluster
            metadata {
              name
              namespace
              selfLink
              creationTimestamp
              annotations
              resourceVersion
              uid
            }
            status
            enforcement
            detail {
              exclude_namespace
              include_namespace
            }
            raw
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves All Policies per Application List Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
        policiesInApplication(
          violatedPolicies:
          [
            {
            name:"policies-policy-mcm-ibm-com-1563995392802",
            namespace:"default",
            clusters:[{name:"cluster1"}]},
            {name:"policy-namespace",namespace:"default",clusters:[{name:"cluster1"}]}
          ]) {
            cluster
            metadata {
              name
              namespace
              selfLink
              creationTimestamp
              annotations
              resourceVersion
              uid
            }
            status
            enforcement
            detail {
              exclude_namespace
              include_namespace
            }
            raw
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Single Policy Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          policies(name:"default.case1-test-policy", clusterName:"calamari") {
            cluster
            message
            metadata {
              name
              namespace
              selfLink
              creationTimestamp
              annotations
              resourceVersion
              uid
            }
            status
            enforcement
            detail {
              exclude_namespace
              include_namespace
            }
            raw
            roleTemplates {
              apiVersion
              complianceType
              compliant
              status
              lastTransition
              name
              kind
              validity
              raw
            }
            roleBindingTemplates {
              apiVersion
              complianceType
              compliant
              status
              lastTransition
              name
              kind
              validity
              raw
            }
            objectTemplates {
              apiVersion
              complianceType
              compliant
              status
              lastTransition
              name
              kind
              validity
              raw
            }
            policyTemplates {
              name
              kind
              lastTransition
              complianceType
              apiVersion
              status
              raw
            }
            violations {
              name
              cluster
              message
              timestamp
            }
            rules {
              complianceType
              templateType
              ruleUID
              verbs
              apiGroups
              resources
            }
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Create Policy Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createPolicy(resources:[{
            apiVersion: "policies.open-cluster-management.io/v1",
            kind: "Policy",
            metadata: {
              name: "test-policy",
              description: "Instance descriptor for policy resource",
            },
            spec: {
              remediationAction: "enforce",
              namespaces: {
                include: [
                  "default",
                ],
                exclude: [
                  "kube*",
                ],
              },
              roleTemplates: [
                {
                  kind: "RoleTemplate",
                  apiVersion: "roletemplate.mcm.ibm.com/v1alpha1",
                  complianceType: "musthave",
                  metadata: {
                    namespace: "",
                    name: "test-role",
                  },
                  selector: {
                    matchLabels: {
                      cloud: "IBM",
                    },
                  },
                  rules: [
                    {
                      complianceType: "musthave",
                      PolicyRule: {
                        apiGroups: [
                          "extensions",
                          "apps",
                        ],
                        resources: [
                          "deployments",
                        ],
                        verbs: [
                          "get",
                          "list",
                          "watch",
                          "delete",
                        ],
                      },
                    },
                    {
                      complianceType: "mustnothave",
                      PolicyRule: {
                        apiGroups: [
                          "core",
                        ],
                        resources: [
                          "pods",
                        ],
                        verbs: [
                          "create",
                          "update",
                          "patch",
                        ],
                      },
                    },
                    {
                      PolicyRule: {
                        apiGroups: [
                          "core",
                        ],
                        resources: [
                          "secrets",
                        ],
                        verbs: [
                          "get",
                          "watch",
                          "list",
                          "create",
                          "delete",
                          "update",
                          "patch",
                        ],
                      },
                    },
                  ],
                },
              ],
            },
          }]),
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Cluster List Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          clustersInPolicy(policy: "policy-namespace", hubNamespace: "mcm") {
            name
            metadata {
              labels
              name
              namespace
              annotations
              uid
              selfLink
            }
            kind
            apiVersion
            spec
            status
            total
            violated
            policy
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Delete Policy Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          deletePolicy(name:"test-policy",namespace:"default")
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
