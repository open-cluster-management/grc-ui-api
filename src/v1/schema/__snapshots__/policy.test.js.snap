// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Resolver Correctly Resolves All Policies per Application List Query 1`] = `
Object {
  "data": Object {
    "policiesInApplication": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "attempts": 5,
          "status": 404,
          "statusCode": 404,
        },
      },
      "locations": Array [
        Object {
          "column": 9,
          "line": 3,
        },
      ],
      "message": "Nock: No match for request {
  \\"method\\": \\"GET\\",
  \\"url\\": \\"http://0.0.0.0/kubernetes/apis/policy.open-cluster-management.io/v1/namespaces/cluster1/policies/\\",
  \\"headers\\": {
    \\"authorization\\": \\"localdev\\",
    \\"host\\": \\"0.0.0.0\\",
    \\"accept\\": \\"application/json\\"
  }
}",
      "path": Array [
        "policiesInApplication",
      ],
    },
  ],
}
`;

exports[`Policy Resolver Correctly Resolves All Policies per Cluster List Query 1`] = `
Object {
  "data": Object {
    "policiesInCluster": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "attempts": 5,
          "status": 404,
          "statusCode": 404,
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Nock: No match for request {
  \\"method\\": \\"GET\\",
  \\"url\\": \\"http://0.0.0.0/kubernetes/apis/policy.open-cluster-management.io/v1/namespaces/cluster1/policies/\\",
  \\"headers\\": {
    \\"authorization\\": \\"localdev\\",
    \\"host\\": \\"0.0.0.0\\",
    \\"accept\\": \\"application/json\\"
  }
}",
      "path": Array [
        "policiesInCluster",
      ],
    },
  ],
}
`;

exports[`Policy Resolver Correctly Resolves Cluster List Query 1`] = `
Object {
  "data": Object {
    "clustersInPolicy": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "attempts": 5,
          "status": 404,
          "statusCode": 404,
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Nock: No match for request {
  \\"method\\": \\"GET\\",
  \\"url\\": \\"http://0.0.0.0/kubernetes/apis/policy.open-cluster-management.io/v1/namespaces/mcm/policies/policy-namespace\\",
  \\"headers\\": {
    \\"authorization\\": \\"localdev\\",
    \\"host\\": \\"0.0.0.0\\",
    \\"accept\\": \\"application/json\\"
  }
}",
      "path": Array [
        "clustersInPolicy",
      ],
    },
  ],
}
`;

exports[`Policy Resolver Correctly Resolves Create Policy Mutation 1`] = `
Object {
  "data": Object {
    "createPolicy": Array [
      Object {},
    ],
  },
}
`;

exports[`Policy Resolver Correctly Resolves Delete Policy Mutation 1`] = `
Object {
  "data": Object {
    "deletePolicy": "test-policy",
  },
}
`;

exports[`Policy Resolver Correctly Resolves Single Policy Query 1`] = `
Object {
  "data": Object {
    "policies": Array [
      Object {
        "cluster": "calamari",
        "detail": Object {
          "exclude_namespace": Array [
            "*",
          ],
          "include_namespace": Array [
            "*",
          ],
        },
        "enforcement": "inform",
        "message": "-",
        "metadata": Object {
          "annotations": Object {
            "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"PR.DS DataSecurity\\",\\"policy.open-cluster-management.io/controls\\":\\"PR.DS-2 Data-in-transit\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST-CSF\\"},\\"name\\":\\"case1-test-policy\\",\\"namespace\\":\\"default\\"},\\"spec\\":{\\"disabled\\":false,\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"TrustedContainerPolicy\\",\\"metadata\\":{\\"name\\":\\"case1-test-policy-trustedcontainerpolicy\\"},\\"spec\\":{\\"imageRegistry\\":\\"quay.io\\",\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"severity\\":\\"low\\"}}}],\\"remediationAction\\":\\"inform\\"}}
",
            "policy.open-cluster-management.io/categories": "PR.DS DataSecurity",
            "policy.open-cluster-management.io/controls": "PR.DS-2 Data-in-transit",
            "policy.open-cluster-management.io/standards": "NIST-CSF",
          },
          "creationTimestamp": "2020-05-17T22:56:03Z",
          "name": "default.case1-test-policy",
          "namespace": "calamari",
          "resourceVersion": "27550274",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/default.case1-test-policy",
          "uid": "fcaf4465-0530-47e5-bbd6-b19c3b28699a",
        },
        "objectTemplates": Array [],
        "policyTemplates": Array [
          Object {
            "apiVersion": "policies.ibm.com/v1alpha1",
            "complianceType": "",
            "kind": "TrustedContainerPolicy",
            "lastTransition": "",
            "name": "case1-test-policy-trustedcontainerpolicy",
            "raw": Object {
              "objectDefinition": Object {
                "apiVersion": "policies.ibm.com/v1alpha1",
                "kind": "TrustedContainerPolicy",
                "metadata": Object {
                  "name": "case1-test-policy-trustedcontainerpolicy",
                },
                "spec": Object {
                  "imageRegistry": "quay.io",
                  "namespaceSelector": Object {
                    "exclude": Array [
                      "kube-system",
                    ],
                    "include": Array [
                      "default",
                    ],
                  },
                  "remediationAction": "inform",
                  "severity": "low",
                },
              },
              "templateType": "policy-templates",
            },
            "status": "Compliant",
          },
        ],
        "raw": Object {
          "apiVersion": "policy.open-cluster-management.io/v1",
          "kind": "Policy",
          "metadata": Object {
            "annotations": Object {
              "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"PR.DS DataSecurity\\",\\"policy.open-cluster-management.io/controls\\":\\"PR.DS-2 Data-in-transit\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST-CSF\\"},\\"name\\":\\"case1-test-policy\\",\\"namespace\\":\\"default\\"},\\"spec\\":{\\"disabled\\":false,\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"TrustedContainerPolicy\\",\\"metadata\\":{\\"name\\":\\"case1-test-policy-trustedcontainerpolicy\\"},\\"spec\\":{\\"imageRegistry\\":\\"quay.io\\",\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"severity\\":\\"low\\"}}}],\\"remediationAction\\":\\"inform\\"}}
",
              "policy.open-cluster-management.io/categories": "PR.DS DataSecurity",
              "policy.open-cluster-management.io/controls": "PR.DS-2 Data-in-transit",
              "policy.open-cluster-management.io/standards": "NIST-CSF",
            },
            "creationTimestamp": "2020-05-17T22:56:03Z",
            "generation": 1,
            "labels": Object {
              "cluster-name": "calamari",
              "cluster-namespace": "calamari",
              "root-policy": "default.case1-test-policy",
            },
            "name": "default.case1-test-policy",
            "namespace": "calamari",
            "ownerReferences": Array [
              Object {
                "apiVersion": "policy.open-cluster-management.io/v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "Policy",
                "name": "case1-test-policy",
                "uid": "e919bf5c-df16-4b90-ae01-a39da7138623",
              },
            ],
            "resourceVersion": "27550274",
            "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/default.case1-test-policy",
            "uid": "fcaf4465-0530-47e5-bbd6-b19c3b28699a",
          },
          "spec": Object {
            "disabled": false,
            "policy-templates": Array [
              Object {
                "objectDefinition": Object {
                  "apiVersion": "policies.ibm.com/v1alpha1",
                  "kind": "TrustedContainerPolicy",
                  "metadata": Object {
                    "name": "case1-test-policy-trustedcontainerpolicy",
                  },
                  "spec": Object {
                    "imageRegistry": "quay.io",
                    "namespaceSelector": Object {
                      "exclude": Array [
                        "kube-system",
                      ],
                      "include": Array [
                        "default",
                      ],
                    },
                    "remediationAction": "inform",
                    "severity": "low",
                  },
                },
              },
            ],
            "remediationAction": "inform",
          },
        },
        "roleBindingTemplates": Array [],
        "roleTemplates": Array [],
        "rules": Array [],
        "status": "unknown",
        "violations": Array [],
      },
    ],
  },
}
`;

exports[`Policy Resolver Correctly Resolves Violation History Query (longer history response) 1`] = `
Object {
  "data": Object {
    "statusHistory": Array [
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-16T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-15T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-14T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-13T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-12T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-11T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-10T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-09T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-08T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-07T19:05:03Z",
      },
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-06T19:05:03Z",
      },
    ],
  },
}
`;

exports[`Policy Resolver Correctly Resolves Violation History Query (no data) 1`] = `
Object {
  "data": Object {
    "statusHistory": Array [],
  },
}
`;

exports[`Policy Resolver Correctly Resolves Violation History Query 1`] = `
Object {
  "data": Object {
    "statusHistory": Array [
      Object {
        "message": "Compliant; notification - pods [sample-nginx-pod] in namespace default exist as specified, therefore this Object template is compliant",
        "timestamp": "2020-09-16T19:05:03Z",
      },
    ],
  },
}
`;
